File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerSpeedSelectLayoutLandscape.java
package net.polyv.android.player.common.ui.component;

import static net.polyv.android.player.sdk.foundation.collections.CollectionsKt.listOf;
import static net.polyv.android.player.sdk.foundation.graphics.ColorsKt.parseColor;
import static net.polyv.android.player.sdk.foundation.graphics.DisplaysKt.dp;
import static net.polyv.android.player.sdk.foundation.lang.PreconditionsKt.requireNotNull;
import static net.polyv.android.player.sdk.foundation.ui.ViewGroupsKt.children;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import net.polyv.android.player.common.R;
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider;
import net.polyv.android.player.common.modules.media.viewmodel.PLVMPMediaViewModel;
import net.polyv.android.player.common.modules.media.viewmodel.viewstate.PLVMPMediaPlayViewState;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerFloatAction;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerViewState;
import net.polyv.android.player.sdk.foundation.collections.PLVSequences;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberStateCompareResult;

import java.util.List;
import java.util.Locale;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * @author Hoshiiro
 */
public class PLVMediaPlayerSpeedSelectLayoutLandscape extends FrameLayout implements View.OnClickListener {

    private static final List<Float> SUPPORT_SPEED_LIST = listOf(0.5F, 1.0F, 1.5F, 2.0F);

    private ImageView speedSelectCloseIv;
    private LinearLayout speedSelectContainer;

    protected Float currentSpeed = null;
    protected boolean isVisible = false;

    public PLVMediaPlayerSpeedSelectLayoutLandscape(@NonNull Context context) {
        super(context);
    }

    public PLVMediaPlayerSpeedSelectLayoutLandscape(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public PLVMediaPlayerSpeedSelectLayoutLandscape(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    {
        LayoutInflater.from(getContext()).inflate(R.layout.plv_media_player_ui_component_speed_select_layout, this);
        speedSelectCloseIv = findViewById(R.id.plv_media_player_speed_select_close_iv);
        speedSelectContainer = findViewById(R.id.plv_media_player_speed_select_container);

        initSpeedSelectLayout();

        setOnClickListener(this);
        speedSelectCloseIv.setOnClickListener(this);
    }

    protected void initSpeedSelectLayout() {
        PLVSequences.wrap(SUPPORT_SPEED_LIST)
                .map(new Function1<Float, TextView>() {
                    @Override
                    public TextView invoke(final Float speed) {
                        TextView tv = new TextView(getContext());
                        tv.setText(String.format(Locale.getDefault(), "%.1fx", speed));
                        tv.setTag(speed);
                        tv.setOnClickListener(new OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                onSelectSpeed(speed);
                            }
                        });
                        return tv;
                    }
                })
                .forEach(new Function1<TextView, Unit>() {
                    @Override
                    public Unit invoke(TextView textView) {
                        MarginLayoutParams lp = new MarginLayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                        lp.topMargin = dp(24).px();
                        lp.bottomMargin = dp(24).px();
                        speedSelectContainer.addView(textView, lp);
                        return null;
                    }
                });
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) return;

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaViewModel.class)
                .getMediaPlayViewState()
                .observeUntilViewDetached(this, new Function1<PLVMPMediaPlayViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPMediaPlayViewState viewState) {
                        currentSpeed = viewState.getSpeed();
                        onViewStateChanged();
                        return null;
                    }
                });

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaControllerViewModel.class)
                .getMediaControllerViewState()
                .observeUntilViewDetached(this, new Function1<PLVMPMediaControllerViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPMediaControllerViewState viewState) {
                        isVisible = viewState.getLastFloatActionLayout() == PLVMPMediaControllerFloatAction.SPEED
                                && !viewState.isMediaStopOverlayVisible();
                        return null;
                    }
                });
    }

    protected void onViewStateChanged() {
        PLVRememberState.rememberStateOf(this, "onUpdateCurrentSpeed")
                .compareLastAndSet(currentSpeed)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        onUpdateCurrentSpeed();
                        return null;
                    }
                });

        PLVRememberState.rememberStateOf(this, "onChangeVisibility")
                .compareLastAndSet(isVisible)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        onChangeVisibility();
                        return null;
                    }
                });
    }

    protected void onUpdateCurrentSpeed() {
        if (currentSpeed == null) {
            return;
        }
        for (View child : children(speedSelectContainer)) {
            if (!(child instanceof TextView)) {
                return;
            }
            if (child.getTag() instanceof Float && ((float) child.getTag()) == currentSpeed) {
                ((TextView) child).setTextColor(parseColor("#3F76FC"));
            } else {
                ((TextView) child).setTextColor(Color.WHITE);
            }
        }
    }

    protected void onChangeVisibility() {
        setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onClick(View v) {
        final int id = v.getId();
        if (id == speedSelectCloseIv.getId()) {
            closeLayout();
        } else if (id == this.getId()) {
            closeLayout();
        }
    }

    private void onSelectSpeed(float speed) {
        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaViewModel.class)
                .setSpeed(speed);
        closeLayout();
    }

    private void closeLayout() {
        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaControllerViewModel.class)
                .popFloatActionLayout();
    }

}

File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerSpeedSelectLayoutLandscape.kt
package net.polyv.android.player.common.ui.component

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import net.polyv.android.player.common.R
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider
import net.polyv.android.player.common.modules.media.viewmodel.PLVMPMediaViewModel
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerFloatAction
import net.polyv.android.player.sdk.foundation.graphics.dp
import net.polyv.android.player.sdk.foundation.graphics.parseColor
import net.polyv.android.player.sdk.foundation.ui.children

/**
 * @author Hoshiiro
 */
private val SUPPORT_SPEED_LIST = listOf("0.5", "0.75", "1", "1.25", "1.5", "2", "3")
private val TEXT_COLOR_SELECTED = parseColor("#3F76FC")
private val TEXT_COLOR_NORMAL = Color.WHITE

class PLVMediaPlayerSpeedSelectLayoutLandscape @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : FrameLayout(context, attrs, defStyle), View.OnClickListener {
    private val speedSelectCloseIv: ImageView
    private val speedSelectContainer: LinearLayout

    init {
        LayoutInflater.from(getContext()).inflate(R.layout.plv_media_player_ui_component_speed_select_layout, this)
        speedSelectCloseIv = findViewById<ImageView>(R.id.plv_media_player_speed_select_close_iv)
        speedSelectContainer = findViewById<LinearLayout>(R.id.plv_media_player_speed_select_container)

        initSpeedSelectLayout()

        setOnClickListener(this)
        speedSelectCloseIv.setOnClickListener(this)
    }

    private fun initSpeedSelectLayout() {
        SUPPORT_SPEED_LIST
            .forEach { speed: String ->
                val textView = TextView(context).apply {
                    text = "${speed}x"
                    setTextColor(TEXT_COLOR_NORMAL)
                    textSize = 12F
                    tag = speed
                    setOnClickListener { onSelectSpeed(speed.toFloat()) }
                }
                val lp = MarginLayoutParams(80.dp().px(), LayoutParams.WRAP_CONTENT)
                lp.topMargin = 24.dp().px()
                lp.bottomMargin = 24.dp().px()
                speedSelectContainer.addView(textView, lp)
            }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) return

        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPMediaViewModel>()
            .mediaPlayViewState
            .observe { viewState ->
                val currentSpeed = viewState.speed
                speedSelectContainer.children()
                    .filter { it is TextView && it.tag is String }
                    .forEach {
                        val textView = it as TextView
                        if ((textView.tag as String).toFloatOrNull() == currentSpeed) {
                            textView.setTextColor(TEXT_COLOR_SELECTED)
                        } else {
                            textView.setTextColor(TEXT_COLOR_NORMAL)
                        }
                    }
            }
            .disposeOnDetached(this)

        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPMediaControllerViewModel>()
            .mediaControllerViewState
            .observe { viewState ->
                val isVisible = viewState.lastFloatActionLayout == PLVMPMediaControllerFloatAction.SPEED
                        && !viewState.isMediaStopOverlayVisible
                visibility = if (isVisible) View.VISIBLE else View.GONE
            }
            .disposeOnDetached(this)
    }

    override fun onClick(v: View) {
        val id = v.id
        if (id == speedSelectCloseIv.id) {
            closeLayout()
        } else if (id == this.id) {
            closeLayout()
        }
    }

    private fun onSelectSpeed(speed: Float) {
        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPMediaViewModel>()
            .setSpeed(speed)
        closeLayout()
    }

    private fun closeLayout() {
        PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
            .get<PLVMPMediaControllerViewModel>()
            .popFloatActionLayout()
    }

}
